README.md
# qrandom

Overview
--------
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/qrandom)](https://cran.r-project.org/package=qrandom)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Travis-CI Build Status](https://travis-ci.com/skoestlmeier/qrandom.svg?branch=master)](https://travis-ci.com/skoestlmeier/qrandom)
[![codecov](https://codecov.io/gh/skoestlmeier/qrandom/branch/master/graph/badge.svg)](https://codecov.io/gh/skoestlmeier/qrandom)
[![Total Downloads](https://cranlogs.r-pkg.org/badges/grand-total/qrandom?color=blue)](https://CRAN.R-project.org/package=qrandom)
[![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html)

`qrandom` is an R package providing an interface to the [ANU Quantum Random Number Generator](https://qrng.anu.edu.au/index.php) provided by the Australian National University. An ultra-high bandwith of true random numbers is generated in real-time by measuring the quantum fluctuations of the vacuum. The quantum Random Number Generator is based on the papers *[Real time demonstration of high bitrate quantum random number generation with coherent laser light](https://doi.org/10.1063/1.3597793)* by Symul et al., (2011) and 
*[Maximization of Extractable Randomness in a Quantum Random-Number Generator](https://doi.org/10.1103/PhysRevApplied.3.054004)* by Haw, et al. (2015).

### Key Features
This package offers functions to retrieve a sequence of random integers, signed integers, hexadecimals, and UUID's. Furthermore, one is able to generate true random samples from a normal or uniform distribution. Functions of `qrandom` are:

* qrandom (sequence of true random numbers)
* qrandomunif (true random numbers from a uniform distribution)
* qrandomnorm (true random numbers from a normal distribution)
* qrandommaxint (true random uniformly distributed signed integers in the range `[-.Machine$integer.max; .Machine$integer.max]`)
* qUUID (true random Universally Unique IDentifiers (UUID), conforming to RFC 4122)

Installation
------------
```r
# The easiest way to install qrandom is to download via CRAN
install.packages("qrandom")

# Alternatively, you can install the development version from GitHub
# install.packages("devtools")
devtools::install_github("skoestlmeier/qrandom")
```

Data
-----
We try our best to provide unique true random numbers. All API requests provided by this package are using SSL. As long as nobody is able to break the encryption protocol, the random numbers you obtain should be unique and secure.

The true random numbers provided by this package are generated in real-time by measuring the quantum fluctuations of the vacuum. The official [QRNG@ANU API](https://qrng.anu.edu.au/contact/api-documentation/) currently supports only a maximum of 1,024 random numbers per request, thus requests for more numbers have to be splitted up into smaller requests of a maximum of 1,024 numbers. In fact, each request may take a couple of seconds to be served.

The greatest possible number of requests per function

* `qrandom(n = 100000, type = "hex16", blocksize = 1024)` takes about 13 minutes and its size is about 201.4 MB

* `qrandomunif(n = 100000)` takes about 7 minutes and its size is about 781.3 KB

* `qrandomnorm(n = 100000, method = "boxmuller")` takes about 8 minutes and its size is about 781.3 KB

* `qrandommaxint(n = 100000)` takes about 4 minutes and its size is about 400.0 KB

* `qUUID(n = 100000)` takes about 4 minutes and its size is about 9.9 MB

via a DSL 16,000 internet connection.

Notes
-----


* **qrandom**

  This function generates a true random sequence of up to 100,000 numbers per request. The type
  - 'uint8' generates integer values between 0 and 255 (both including).
  - 'uint16' generates integer values between 0 and 65,535 (both including).
  - 'hex16' generates hexadecimal values between 00 and ff (both including).
  
  The option 'blocksize' is only needed for request type 'hex16' and sets the length of each block which can be a length between 1 and 1,024 (both including).
  Further information can be obtained by the official QRNG\@ANU API documentation [here](https://qrng.anu.edu.au/contact/api-documentation/).

* **qrandomunif**

  This function returns a sample of 1 - 100,000 true random numbers from a uniform distribution with parameters a (minimum value) and b (maximum value), both a and b included, i.e. all values are within the interval [a; b]. Per default (a=0 and b=1), a standard uniform distribution is assumed.


* **qrandomnorm**

  This function returns a sample of 1 - 100,000 true random numbers from a normal distribution. Per default, a standard normal distribution with mean zero and standard deviation of one is assumed.
  
  Internally, uniformly distributed true random numbers within the interval [0; 1] are requested via `qrandomunif()`. Within these uniformly data, the smallest possible number greater than zero is 2.220446e-16 and the largest possible number less than one is 0.9999999999999997779554.

  We provide three methods to transform our standard uniformly data into a normal distribution:

  * [Inverse transform sampling](https://en.wikipedia.org/w/index.php?title=Inverse_transform_sampling&oldid=866923287): The sample of standard uniformly data is interpreted as a probability and transformed into a normal distribution applying the `stats::qnorm()` function.

  * The [polar-method](https://en.wikipedia.org/w/index.php?title=Marsaglia_polar_method&oldid=871161902) by George Marsaglia.

  * [Box-Muller transformation](https://en.wikipedia.org/w/index.php?title=Box-Muller_transform&oldid=873905617) by George Box and Mervin Muller.


  Be aware that only the default method 'inverse' is able to return -Inf and +Inf z-values for the normal distribution. The following table summarizes the non-infinite minimum and maximum z-values for a standard normal distribution for each method provided and compares them with the non-infinite extreme values from the R-core function `stats::qnorm()`:
  
  | method 				| stats:qnorm()	| inverse | polar    | boxmuller |
  | ----------------- | -----------	| --------| ---------| ----------|
  | minimum z-value* 	| -8.209536 	| -8.12589| -8.36707 | -8.490424 |
  | maximum z-value* 	| 8.209536 	| 8.12589 | 8.36707  | 8.490424  |
  | z-values +- Inf 	| Yes 			| Yes     | No       | No        |
  
  **non-infinite values*.



* **qrandommaxint**

  This function returns a sample of 1 - 100,000 true random uniformly distributed signed integers in the range `[-.Machine$integer.max; .Machine$integer.max]`.

* **qUUID**

  This function returns true random Universally Unique IDentifiers (UUID), conforming to RFC 4122.


To-Do
------------
* Add more tests to increase codecov.
* Ziggurat algorithm for 'qrandomnorm'.
* Speed up / parallelize requests for thousands of true random numbers in 'qrandom'.

Contributing
------------
Constributions in form of feedback, comments, code, bug reports or pull requests are most welcome. How to contribute:

* Issues, bug reports, or desired expansions: File a GitHub issue.
* Fork the source code, modify it, and issue a pull request through the project GitHub page.

Please read the [contribution guidelines](CONTRIBUTING.md) on how to contribute to this R-package.

Code of conduct
------------
Please note that this project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.

Legal issues
------------
`qrandom` -- An R interface to the ANU Quantum Random Numbers Server

Copyright (C) 2020  Siegfried Köstlmeier <siegfried.koestlmeier@gmail.com>

`qrandom` is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

`qrandom` is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with `qrandom`. If not, see <http://www.gnu.org/licenses/>.


NEWS.md
# qrandom 1.2.6
* Bug fixes for v.1.2.6.

# qrandom 1.2.5
* Manual adjustments for KaTeX (or Mathjax) to render LaTeX-like mathematical equations.

# qrandom 1.2.4
* Bug fixes for v.1.2.3

# qrandom 1.2.3
* This update implements the CRAN policy for the provided functions to fail gracefully if no internet resource is available:
'Packages which use Internet resources should fail gracefully with an informative message if the resource is not available (and not give a check warning nor error)'.

# qrandom 1.2.2
* This update extends the implementation of v1.2.1. While the latter just checks if the device is connected to the internet, this version explicitely checks if the ANU Quantum Random Number Generator provided by the Australian National University is available and accessible by its provided API.

# qrandom 1.2.1
* The CRAN policy states that packages which use Internet resources should fail gracefully with an informative message if the resource is not available (and not give a check warning nor error). This is implemented both in the package functions and the testthat unit tests.

# qrandom 1.2
* Added `testthat` test-cases for further development

# qrandom 1.1
## New features

* qrandommaxint (true random signed integers)
* qUUID (true random Universally Unique IDs (UUID))

# qrandom 1.0
## New features

`qrandom` implements an interface to the ANU Quantum Random Numbers Server and offers true random numbers by measuring the quantum fluctuations of the vacuum.

This package offers functions to retrieve a sequenze of random integers or hexadecimals and to generate true random samples from a normal or uniform distribution. Functions of `qrandom` are:

* qrandom (sequence of true random numbers)
* qrandomunif (true random numbers from a uniform distribution)
* qrandomnorm (true random numbers from a normal distribution)

## References
  Secure Quantum Communication group,
  Centre for Quantum Computing and Communication Technology,
  Department of Quantum Sciences,
  Research School of Physics and Engineering,
  The Australian National University, Canberra, ACT 0200, Australia.
  *Welcome to the ANU Quantum Random Numbers Server.*
  https://qrng.anu.edu.au/index.php
  
  T. Symul, S. M. Assad and P. K. Lam (2011):
  Real time demonstration of high bitrate quantum random number generation with coherent laser light .
  *Applied Physics Letters*, **98**, 231103.
  doi: [10.1063/1.3597793](https://doi.org/10.1063/1.3597793).
  
  J. Y. Haw, S. M. Assad, A. M. Lance, N. H. Y. Ng, V. Sharma, P. K. Lam, and T. Symul (2015):
  Maximization of Extractable Randomness in a Quantum Random-Number Generator.
  *Physical Review Applied*, **3**, 054004.
  doi: [10.1103/PhysRevApplied.3.054004](https://doi.org/10.1103/PhysRevApplied.3.054004).


tests/testthat.R
## load dependencies
library(testthat)
library(qrandom)

check_qrng <- function(){
  tryCatch(
    expr = {
      req <- curl::curl_fetch_memory('https://qrng.anu.edu.au/')
      req$status_code
    },
    error = function(e){
      -1
    }
  )
}

## test package separated with filter due to limited Travis-CI build time
if(curl::has_internet() && check_qrng() == 200){
  test_check('qrandom', filter = "qrandom")
  test_check('qrandom', filter = "qrandomunif")
  test_check('qrandom', filter = "qrandomnorm")
  test_check('qrandom', filter = "qUUID")
  test_check('qrandom', filter = "qrandommaxint")
}


tests/testthat/test-qUUID.R
context('functions')

test_that("qUUID() returns valid output", {
  skip_on_cran()

  x <- qUUID(10)

  # returns values
  expect_equal(10, length(x))
  # correct string length
  expect_true(all(nchar(x) == 36))

  x <- gsub("-", "", x)   # purge hyphens
  expect_true(all(nchar(x) == 32))

  # test valid characters
  expect_true(all(! grepl("[^0-9a-f]", x)))

  x <- sapply(x, FUN = x32bit)    # convert to bit matrix

  expect_true(all(x[61, ] == 0))
  expect_true(all(x[62, ] == 1))
  expect_true(all(x[63, ] == 0))
  expect_true(all(x[64, ] == 0))

  expect_true(all(x[71, ] == 0))
  expect_true(all(x[72, ] == 1))

})

# [END]


tests/testthat/test-qrandom.R
context('functions')

test_that('qrandom', {
  skip_on_cran()

  tmp <- qrandom()

  expect_equal(length(tmp), 1)
  expect_success(expect_error(qrandom(n = "a"), "The number 'n' of random numbers to return has to be an integer."))
  expect_success(expect_error(qrandom(n = 2.5, blocksize = 1), "The number 'n' of random numbers to return has to be an integer."))
  expect_success(expect_error(qrandom(n = -3), "Random number requests must be between 1 and 100,000 numbers."))
  expect_success(expect_error(qrandom(n = 100001), "Random number requests must be between 1 and 100,000 numbers."))
  expect_failure(expect_error(qrandom(type = "uint8"), "Type has to be 'uint8', 'uint16' or 'hex16'."))
  expect_success(expect_error(qrandom(type = "INVALID-TYPE"), "Type has to be 'uint8', 'uint16' or 'hex16'."))
  expect_failure(expect_error(qrandom(blocksize =  1), "The variable 'blocksize' has to be an integer."))
  expect_success(expect_error(qrandom(blocksize =  1.5), "The variable 'blocksize' has to be an integer."))
  expect_success(expect_error(qrandom(blocksize = "a"), "The variable 'blocksize' has to be an integer."))
  expect_success(expect_error(qrandom(blocksize = 0), "The variable 'blocksize' must be between 1 and 1,024."))
  expect_success(expect_error(qrandom(blocksize = 1025), "The variable 'blocksize' must be between 1 and 1,024."))
  expect_success(expect_error(qrandom(blocksize = 2.5), "The variable 'blocksize' has to be an integer."))
  expect_equal(class(tmp), c("integer"))
  expect_equal(dim(tmp), NULL)

  tmp<-qrandom(type = "hex16")
  expect_equal(class(tmp), c("character"))
  expect_equal(dim(tmp), NULL)

  tmp <- qrandomnorm(method = "polar")
  expect_type(tmp, "double")

  tmp <- qrandomnorm(method = "boxmuller")
  expect_type(tmp, "double")

})


R/qUUID.R
##  qUUID -- quantum random UUIDs.
##
##  Copyright (C) 2019  Boris Steipe <boris.steipe@utoronto.ca>
##                      ORCID: 0000-0002-1134-6758
##
##  This file is part of the qrandom R-package
##
##  qrandom is free software; you can redistribute it and/or
##  modify it under the terms of the GNU General Public License
##  as published by the Free Software Foundation; either version 2
##  of the License, or (at your option) any later version.
##
##  qrandom is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with qrandom. If not, see <http://www.gnu.org/licenses/>.


x32bit <- function(x) {
  # x: a 32 digit hex string
  # return: a 128 raw vector, its binary representation
  idx <- seq(1, 32, by = 4)
  x8 <- substring(x, idx, idx + 3)
  x8 <- paste0("0x", x8)
  s <- c(13:16, 9:12, 5:8, 1:4)  # undo reversal by strtoi
  m8 <- sapply(x8, FUN = function(x) { intToBits(strtoi(x))[s] })
  return(as.vector(m8))
}

stampUUIDversion <- function(b) {
  # stamp the UUID random version identifier onto a 128 bit raw vector.
  # b: a length 128 raw vector
  # return: a length 128 raw vector with a RFC 4122 conforming version pattern
  b[61:64] <- as.raw(c(0, 1, 0, 0))
  b[71:72] <- as.raw(c(0, 1))
  return(b)
}

bit128x <- function(b) {
  # b: a length 128 raw vector
  # return: the 32 digit hex string represented by b
  dim(b) <- c(4, 32)
  hMap <- c("0", "1", "2", "3", "4", "5", "6", "7",
            "8", "9", "a", "b", "c", "d", "e", "f")
  p2 <- 2^(0:3)
  x <- apply(b, 2, FUN = function(x) {hMap[sum(as.integer(x) * p2) + 1]} )
  return(paste0(x, collapse = ""))
}

hyphenate <- function(x) {
  # x: a 32 character string
  # value: a 36 character string hyphenated in a 8-4-4-4-12 pattern
  from <- c(1, 9, 13, 17, 21)
  to <- c(8, 12, 16, 20, 32)
  return(paste(substring(x, from, to), collapse = "-"))
}

qUUID <- function(n = 1) {
  # n: number of uuids to return
  # value: n RFC 1422 conforming, random UUIDs

  if(!curl::has_internet()){
    message("No Internet connection! \n")
    return(invisible(NULL))
  }

  if(!check_qrng() == 200){
    message("The ANU Quantum Random Number Generator service is currently not available at [https://qrng.anu.edu.au/index.php]. \n")
    return(invisible(NULL))
  }else{

  x <- qrandom::qrandom(n = n, type = "hex16", blocksize = 16)
  x <- tolower(x)
  x <- sapply(x, FUN = x32bit)
  x <- apply(x, 2, FUN = stampUUIDversion)
  x <- apply(x, 2, FUN = bit128x)
  x <- sapply(x, FUN = hyphenate)
  names(x) <- NULL

  return(x)
  }
}

# [END]


R/qrandommaxint.R
##  qrandommaxint -- quantom random integers from the full range of signed
##                   integers.
##
##  Copyright (C) 2019  Boris Steipe <boris.steipe@utoronto.ca>
##                      ORCID: 0000-0002-1134-6758
##
##  This file is part of the qrandom R-package
##
##  qrandom is free software; you can redistribute it and/or
##  modify it under the terms of the GNU General Public License
##  as published by the Free Software Foundation; either version 2
##  of the License, or (at your option) any later version.
##
##  qrandom is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with qrandom. If not, see <http://www.gnu.org/licenses/>.


x2bit <- function(x) {
  # x: an 8 digit hex string
  # return: a 32 integer vector of [0, 1] as its binary representation
  x4 <- substring(x, c(1, 3, 5, 7), c(2, 4, 6, 8))
  x4 <- rev(paste0("0x", x4))
  m4 <- sapply(x4, FUN = function(x) { intToBits(strtoi(x))[1:8] })
  return(as.integer(as.vector(m4)))
}

bit2int <- function(x) {
  # x: a 32 integer vector of [0, 1]
  # return: its signed integer equivalent
  p2 <- 2^(0:30)                              # powers of two
  s <- 1                                      # initialize sign
  if (x[32] == 1) {                           # sign bit is set
    x <- c(ifelse(x[1:31] == 1, 0, 1), x[32]) # flip bits
    s <- -1                                   # sign is negative
  }
  return(as.integer((s * sum(x[1:31] * p2)) - x[32]))
}

qrandommaxint <- function(n = 1) {
  # n: integer: number of values to return. See restrictions of qrandom().
  #             Default 1.
  # return: integer: n signed 32-bit integers in the range
  #                  [-.Machine$integer.max; .Machine$integer.max]

  if(!curl::has_internet()){
    message("No Internet connection! \n")
    return(invisible(NULL))
  }

  if(!check_qrng() == 200){
    message("The ANU Quantum Random Number Generator service is currently not available at [https://qrng.anu.edu.au/index.php]. \n")
    return(invisible(NULL))
  }else{

  x <- qrandom::qrandom(n = n, type = "hex16", blocksize = 4)
  x[x == "80000000"] <- "00000000"  # input would underflow: "-0"

  int <- sapply(x, FUN = function(x) { bit2int(x2bit(x)) }, USE.NAMES = FALSE)

  return(int)
  }
}

# [END]


R/qrandom.R
##  qrandom -- An R interface to the ANU Quantum Random Numbers Server
##
##  Copyright (C) 2018  Siegfried Köstlmeier <siegfried.koestlmeier@gmail.com>
##
##  This file is part of the qrandom R-package
##
##  qrandom is free software; you can redistribute it and/or
##  modify it under the terms of the GNU General Public License
##  as published by the Free Software Foundation; either version 2
##  of the License, or (at your option) any later version.
##
##  qrandom is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with qrandom. If not, see <http://www.gnu.org/licenses/>.

check_qrng <- function(){
  tryCatch(
    expr = {
      req <- curl::curl_fetch_memory('https://qrng.anu.edu.au/')
      req$status_code
    },
    error = function(e){
      -1
    }
  )
}

getConnection <- function(website, ...) {
  if (capabilities()["libcurl"]) {
    url(website, ..., method = "libcurl")
  } else {
    curl(website, ...)
  }
}

closeConnection <- function(con) {
  if (capabilities()["libcurl"]) {
    close(con)
  }
}


get_sequence <- function(n = 1,
                         type = "uint8",
                         blocksize = 1) {
  urlbase <- "https://qrng.anu.edu.au/API/jsonI.php"

  urltxt <- paste(
    urlbase,
    "?length=",
    format(n, scientific = FALSE),
    "&type=",
    format(type, scientific = FALSE),
    sep = ""
  )

  if (type == "hex16") {
    urltxt <- paste(urltxt,
                    "&size=",
                    format(blocksize, scientific = FALSE),
                    sep = "")
  }

  con <- getConnection(website = urltxt, open = "r")
  rt <- readLines(con, encoding = "UTF-8")
  randNum <- as.character(rt)
  randNum <- fromJSON(randNum)
  randNum <- randNum$data
  closeConnection(con)
  return(randNum)
}

qrandom <- function(n = 1,
                    type = "uint8",
                    blocksize = 1) {
  if (!is.numeric(n) || !n %% 1 == 0) {
    stop("The number 'n' of random numbers to return has to be an integer.")
  }
  if (n < 1 || n > 100000) {
    stop("Random number requests must be between 1 and 100,000 numbers.")
  }
  if (!type %in% c("uint8", "uint16", "hex16")) {
    stop("Type has to be 'uint8', 'uint16' or 'hex16'.")
  }
  if (!is.numeric(blocksize) || !blocksize %% 1 == 0) {
    stop("The variable 'blocksize' has to be an integer.")
  }
  if (blocksize < 1 || blocksize > 1024) {
    stop("The variable 'blocksize' must be between 1 and 1,024.")
  }

  if(!curl::has_internet()){
    message("No Internet connection! \n")
    return(invisible(NULL))
  }

  if(!check_qrng() == 200){
    message("The ANU Quantum Random Number Generator service is currently not available at [https://qrng.anu.edu.au/index.php]. \n")
    return(invisible(NULL))
  }
  else{

  tmp <- c()

  count <- n %/% 1024
  remain <- n %% 1024

  if (count != 0) {
    ##  true, if more than 1,024 numbers are requested
    progress <- txtProgressBar(min = 0,
                               max = count,
                               style = 3)
    for (i in 1:count) {
      tmp <-
        c(tmp,
          get_sequence(
            n = 1024,
            type = type,
            blocksize = blocksize
          ))
      setTxtProgressBar(progress, i)
    }
    if (remain != 0) {
      tmp <-
        c(tmp,
          get_sequence(
            n = remain,
            type = type,
            blocksize = blocksize
          ))
    }
    close(progress)
  } else{
    tmp <-
      get_sequence(n = n,
                   type = type,
                   blocksize = blocksize)
  }
  return(tmp)
  }
}

qrandomunif <- function(n = 1,
                        a = 0,
                        b = 1) {
  if (!is.numeric(n) || !n %% 1 == 0) {
    stop("The number 'n' of random numbers to return has to be an integer.")
  }
  if (n < 1 || n > 100000) {
    stop("Random number requests must be between 1 and 100,000 numbers.")
  }
  if (!is.numeric(a) || !is.numeric(b)) {
    stop("The parameters 'a' and 'b' must be numeric.")
  }
  if (!a < b) {
    stop("The minimum value 'a' of the distribution has to be less than parameter 'b'.")
  }

  ## We use this function qrandomunif() for further calculations in qrandomnorm(), so
  ## we decided to use hexadecimal numbers with block-size 7 for uniform distribution data
  ## and deleting the first character of each hexadecimal number.
  ## Therefore, the numbers here are uniformly distributed within the range [0x0000000000000; 0xfffffffffffff]
  ## which is [0; 4,503,599,627,370,495] in decimal integers.
  ## As we already have uniformly distributed numbers, we just divide each number by the maximum possible
  ## value of 0xfffffffffffff to normalize our random numbers. After conversion into decimal numbers we
  ## obtain uniformly distributed numbers within the interval [0; 1].
  ##
  ## See the comment on qrandomnorm() for futher information on why we choose especially
  ## block-size 7 and delete the first character of each hexadecimal number.

  if(!curl::has_internet()){
    message("No Internet connection! \n")
    return(invisible(NULL))
  }

  if(!check_qrng() == 200){
    message("The ANU Quantum Random Number Generator service is currently not available at [https://qrng.anu.edu.au/index.php]. \n")
    return(invisible(NULL))
  }else{

  tmp <-
    qrandom(n = n,
            type = "hex16",
            blocksize = 7)

  # cast 'a' and 'b' to mpfr
  # see https://github.com/skoestlmeier/qrandom/issues/2 for further details
  a_tmp <- mpfr(a, base = 10, precBits = 52)
  b_tmp <- mpfr(b, base = 10, precBits = 52)

  ## delete first (random) character
  tmp <- substring(tmp, 2)

  tmp <- mpfr(tmp, base = 16)
  tmp <- new("mpfr", unlist(tmp))

  ## Normalization within the interval [0; 1]
  urand <- tmp / mpfr("fffffffffffff", base = 16)
  #urand <- as.numeric(urand)


  if (a != 0 || b != 1) {
    ## transform distribution to have minimum value 'a' and maximum value of 'b',
    ## so data is uniformly distributed within the interval [a, b].
    urand <- (b_tmp - a_tmp) * urand + a_tmp
  }

  return(as.numeric(urand))
  }
}

qrandomnorm <- function(n = 1,
                        mean = 0,
                        sd = 1,
                        method = "inverse") {
  if (!is.numeric(n) || !n %% 1 == 0) {
    stop("The number 'n' of random numbers to return has to be an integer.")
  }
  if (n < 1 || n > 100000) {
    stop("Random number requests must be between 1 and 100,000 numbers.")
  }
  if (!is.numeric(mean) || !is.numeric(sd)) {
    stop("The parameters 'mean' and 'sd' must be numeric.")
  }
  if (sd < 0) {
    stop("The standard deviation 'sd' must not be negative.")
  }
  if (!method %in% c("inverse", "polar", "boxmuller")) {
    stop(
      "The method for generating true normal random variables has to be 'inverse', 'polar' or 'boxmuller'."
    )
  }

  if(!curl::has_internet()){
    message("No Internet connection! \n")
    return(invisible(NULL))
  }

  if(!check_qrng() == 200){
    message("The ANU Quantum Random Number Generator service is currently not available at [https://qrng.anu.edu.au/index.php]. \n")
    return(invisible(NULL))
  }else{

  ## Our procedure described in the initial comment on qrandomunif() garanties that qrandomunif() returns true random numbers,
  ## where the smallest possible number greater than zero is 2.220446e-16 and the largest possible number less than
  ## one is 0.9999999999999997779554. This is useful as the function qrandomnorm() is internally based on the uniformly data in qrandomunif().

  x <- vector("numeric", n)

  if (method == "inverse") {
    ## Using qrandomunif() for getting standard uniformly distributed true random numbers
    ## and applying the Inverse transform sampling described in https://en.wikipedia.org/w/index.php?title=Inverse_transform_sampling&oldid=866923287
    ## (which is also known as inversion sampling, the inverse probability integral transform, or the inverse transformation method).

    tmp <- qrandomunif(n = n)

    ## tmp is within the range [0; 1]. Applying qnrom() results in -Inf if tmp equals 0 and +Inf if tmp equals 1.
    ## Applying stats::qnorm() on the smallest possible number greater than zero (2.220446e-16) and on the
    ## largest possible number less than one (0.9999999999999997779554) results in a limitation for our
    ## non-infinite z-values within the interval [-8.125891; 8.125891]. In comparison, the non-infinite z-values
    ## for the function stats::qnorm() are limited within the interval [-8.209536; 8.209536].

    ## +-Inf for z-values of normal distributed data are only possible when using this 'inverse' method.
    tmp <- qnorm(tmp)
    x <- sd * tmp + mean
  } else if (method == "polar") {
    ## Only non-infinite z-values are possible when using this 'polar' method.
    ## Apply method 'inverse' for the possibilty of +-Inf z-values.
    ## z-values for a standard normal distribution with mean = 0 and sd = 1 are bound within the
    ## interval [-8.36707;  8.36707].
    x <- marsaglia_bray(n = n, mean = mean, sd = sd)
  } else if (method == "boxmuller") {
    ## Only non-infinite z-values are possible when using this 'boxmuller' method.
    ## Apply method 'inverse' for the possibilty of +-Inf z-values.
    ## z-values for a standard normal distribution with mean = 0 and sd = 1 are bound within the
    ## interval [-8.490424;  8.490424]
    x <- box_muller(n = n, mean = mean, sd = sd)
  }

  return(x)
  }
}

## Implementation of the Marsaglia Bray Polar method, see https://en.wikipedia.org/w/index.php?title=Marsaglia_polar_method&oldid=871161902
## and https://stackoverflow.com/a/43619239/8512077 for the original code
marsaglia_bray <- function(n = 1,
                           mean = 0,
                           sd = 1)
{
  x <- vector("numeric", n)

  i <- 1

  count <- 1
  cat("Request 1/2:\n")
  tmp1 <- qrandomunif(n)
  cat("Request 2/2:\n")
  tmp2 <- qrandomunif(n)

  while (i <= n)
  {
    if (count > n) {
      count <- 1
      tmp1 <- qrandomunif(n)
      tmp2 <- qrandomunif(n)
    }
    u <- c(tmp1[count], tmp2[count])
    v <- 2 * u - 1
    w <- sum(v ^ 2)

    if (w > 0 && w < 1)
    {
      y <- sqrt(-2 * log(w) / w)

      z <- v * y

      x[i] <- z[1]

      if ((i + 1) <= n)
      {
        x[i + 1] <- z[2]
        i <- i + 1
      }

      i <- i + 1
    }
    count <- count + 1
  }

  x * sd + mean
}

## Implementation of the Box–Muller transformation, see https://en.wikipedia.org/w/index.php?title=Box%E2%80%93Muller_transform&oldid=873905617
## and https://stackoverflow.com/a/43619239/8512077 for the original code
box_muller <- function(n = 1,
                       mean = 0,
                       sd = 1)
{
  x <- vector("numeric", n)

  count <- 1
  cat("Request 1/2:\n")
  tmp1 <- qrandomunif(n)
  cat("Request 2/2:\n")
  tmp2 <- qrandomunif(n)

  i <- 1
  while (i <= n)
  {
    if (count > n) {
      count <- 1
      tmp1 <- qrandomunif(n)
      tmp2 <- qrandomunif(n)
    }

    u1 <- tmp1[count]
    u2 <- tmp2[count]
    if (!u1 == 0) {
      x[i] <- sqrt(-2 * log(u1)) * cos(2 * pi * u2)

      if ((i + 1) <= n)
      {
        x[i + 1] <- sqrt(-2 * log(u1)) * sin(2 * pi * u2)
        i <- i + 1
      }

      i <- i + 1
    }
    count <- count + 1
  }

  x * sd + mean
}
